File: ESP32/Persistence.py
import time
import uos as os
import io


class Persistence:
    def __init__(self):
        dataFileName = "sensorData.txt"
        if dataFileName not in os.listdir():
            x = open(dataFileName, 'x')
            x.close()
        print("Initialized Persistence.__init__()")

    def sensorDataToBytes(self, temp: float, humidity: float) -> bytes:
        # Byte-Composition for Sensor-Data:
        # AAAABBBB | BCCCCCCC
        # A -> 4Bit for Temp. comma -> 0-9 / 16
        # B -> 5Bit for Temp. dezimal -> 0-32
        # C -> 7Bit for Humidity -> 0-100 / 128

        tDec = int(str(temp).split('.')[0])
        tComma = int(str(temp).split('.')[1][:1])
        humByte = int(humidity)

        decLSB = tDec & 1
        decRest = (tDec >> 1) & 0b00001111
        humCut = humByte & 0b01111111

        firstByte = (tComma << 4) | decRest
        secondByte = (decLSB << 7) | humCut

        return int(firstByte).to_bytes(1, 'big') + int(secondByte).to_bytes(1, 'big')

    def persistedBytesToValues(self, persistedBytes: bytes) -> [float, int]:
        firstByte = int(persistedBytes[0])
        secondByte = int(persistedBytes[1])

        tempComma = (firstByte >> 4) / 10
        tempDec = (secondByte >> 7) | ((firstByte & 0b00001111) << 1)
        humidity = secondByte & 0b01111111

        return [tempDec + tempComma, humidity]

    async def persistValue(self, temperature: float, humidity: float, currDateMEZ: time):
        print(currDateMEZ)

        dataFile = open('sensorData.txt', 'ab+')

        matchOrder = ['yy:'.encode() + int(currDateMEZ[0]).to_bytes(2, 'big'),
                      'mm:'.encode() + int(currDateMEZ[1]).to_bytes(2, 'big'),
                      'dd:'.encode() + int(currDateMEZ[2]).to_bytes(2, 'big'),
                      'hh:'.encode() + int(currDateMEZ[3]).to_bytes(2, 'big')]

        while True:
            line = dataFile.read(5)
            if line.startswith(matchOrder[0]):
                break

            if line == ''.encode():
                dataFile.write('yy:'.encode() + int(currDateMEZ[0]).to_bytes(2, 'big'))
                print('Created Year entry')
                break

        while True:
            line = dataFile.read(5)
            if line.startswith(matchOrder[1]):
                break

            if line == ''.encode():
                dataFile.write('mm:'.encode() + int(currDateMEZ[1]).to_bytes(2, 'big'))
                print('Created Month entry')
                break

        while True:
            line = dataFile.read(5)
            if line.startswith(matchOrder[2]):
                break

            if line == ''.encode():
                dataFile.write('dd:'.encode() + int(currDateMEZ[2]).to_bytes(2, 'big'))
                print('Created Day entry')
                break

        while True:
            line = dataFile.read(5)
            if line.startswith(matchOrder[3]):
                break

            if line == ''.encode():
                dataFile.write('hh:'.encode() + int(currDateMEZ[3]).to_bytes(2, 'big'))
                print('Created Hour entry')
                break

        minuteEntryExists = False
        while True:
            line = dataFile.read(5)
            if line.startswith(int(currDateMEZ[4]).to_bytes(2, 'big') + ':'.encode()):
                minuteEntryExists = True
                break

            if line == ''.encode():
                break

        # Now we know that the Data-Structure exists -> save Data
        if not minuteEntryExists:
            data = self.sensorDataToBytes(temperature, humidity)
            dataFile.write(int(currDateMEZ[4]).to_bytes(2, 'big'))
            dataFile.write(':'.encode() + data)

        dataFile.close()

    def getDataByTimestamp(self, timestamp: []):
        dataFile = open('sensorData.txt', 'rb')

        matchOrder = ['yy:'.encode() + int(timestamp[0]).to_bytes(2, 'big'),
                      'mm:'.encode() + int(timestamp[1]).to_bytes(2, 'big'),
                      'dd:'.encode() + int(timestamp[2]).to_bytes(2, 'big'),
                      'hh:'.encode() + int(timestamp[3]).to_bytes(2, 'big')]

        dataTime = int(timestamp[4] / 5) * 5
        data = ''

        currentMtcIndex = 0

        line = dataFile.read(5)
        while data == '' and line[0] != ''.encode():
            if line.__contains__(matchOrder[currentMtcIndex]) and currentMtcIndex <= 3:
                currentMtcIndex += 1

            if currentMtcIndex == 4:
                if line.startswith(dataTime.to_bytes(2, 'big')):
                    data = line

            line = dataFile.read(5)

        dataFile.close()
        if data != '':
            return self.persistedBytesToValues(data.split(':'.encode(), 1)[1])

        return [0, 0]

    def decodeWholeFile(self):
        dataFile = open('sensorData.txt', 'rb')

        body = ''
        while True:
            line = dataFile.read(5).split(':'.encode())
            line_start = line[0].decode()

            if line_start == '':
                break

            if line_start.startswith('yy') or line_start.startswith('mm') \
                    or line_start.startswith('dd') or line_start.startswith('hh'):
                body = body + line_start + ':' + str(int.from_bytes(line[1], 'big')) + '\n'
            else:
                body = body + str(int.from_bytes(line[0], 'big')) + ':' + str(
                    self.persistedBytesToValues(line[1])) + '\n'

        dataFile.close()
        return body

    def overrideFile(self, data):
        dataFile = open('sensorData.txt', 'wb')

        buf = io.StringIO(data)
        while True:
            line = buf.readline().strip()

            if line == '':
                break

            line = line.split(':')
            if line[0].startswith('yy') or line[0].startswith('mm') or line[0].startswith('dd') or line[0].startswith(
                    'hh'):
                dataFile.write(line[0].encode() + ':'.encode() + int(line[1].strip()).to_bytes(2, 'big'))
            else:
                sensorData = line[1].replace('[', '').replace(']', '').replace(' ', '').split(',')
                dataFile.write((int(line[0]).to_bytes(2, 'big') + ':'.encode() + self.sensorDataToBytes(
                    float(sensorData[0]), int(sensorData[1]))))

        dataFile.close()
        print('Overwritten sensorData.txt')

    def getCurrentMEZ(self):
        # Austria-MESZ between:
        # Last March-Sunday at 02:00 MEZ
        # Last Oktober-Sunday at 03:00 MESZ
        mez = (time.time() + 3600)
        t = time.localtime(mez)
        month = t[1]
        day = t[2]

        if 3 <= month <= 10:
            if month == 3:
                lastSunday = self.getLastSundayOfMonth(mez)
                if day > lastSunday or (day == lastSunday and t[3] >= 2):
                    mez = mez + 3600
            elif month == 10:
                mest = mez + 3600
                tmest = time.localtime(mest)
                lastSunday = self.getLastSundayOfMonth(mest)
                if tmest[2] < lastSunday or (tmest[2] == lastSunday and tmest[3] < 3):
                    mez = mez + 3600
            else:
                mez = mez + 3600

        return mez

    def getLastSundayOfMonth(self, mez) -> int:
        t = time.localtime(mez)
        currDay = t[2]
        diff = 0

        if t[1] == 3 or t[1] == 10:
            diff = 31 - currDay

        eomTime = mez + (3600 * 24 * diff)
        eom = time.localtime(eomTime)
        weekDay = eom[6]
        diffToPrevSunday = 7 - (6 - weekDay)

        lastSunday = eom[2]
        if diffToPrevSunday < 7:
            lastSunday = time.localtime(eomTime - (3600 * 24 * diffToPrevSunday))[2]

        return lastSunday

File: ESP32_cpp/src/Persistence.cpp
#include <Persistence.h>
#include <cmath>
#include <sstream>
#include <iostream>

Persistence Storage;

uint16_t Persistence::sensorDataToBytes(float tmp, float hum) {
        // Byte-Composition for Sensor-Data:
        // AAAABBBB | BCCCCCCC
        // A -> 4Bit for Temp. comma -> 0-9 / 16
        // B -> 5Bit for Temp. dezimal -> 0-32
        // C -> 7Bit for Humidity -> 0-100 / 128
        uint8_t tmp_dec = std::floor(tmp);
        uint8_t tmp_comma = ( (tmp - tmp_dec)) * 10 + 0.5f;
        uint8_t hum_dec = std::floor(hum);

        uint8_t decLSB = tmp_dec & 1;
        uint8_t decRest = (tmp_dec >> 1) & 0b00001111;
        uint8_t hum_byte = hum_dec & 0b01111111;
        
        uint8_t firstByte = (tmp_comma << 4) | decRest;
        uint8_t secondByte = (decLSB << 7) | hum_byte;

        uint16_t bytes_to_persist = firstByte << 8 | secondByte;
        return bytes_to_persist;
}

void Persistence::persistedBytesToValue(uint8_t firstByte, uint8_t secondByte, float* ret) {
        float tmp_comma = ((float)(firstByte >> 4)) / 10;

        uint8_t tmp_dec = (secondByte >> 7) | ((firstByte & 0b00001111) << 1);
        uint8_t humidity = secondByte & 0b01111111;

        ret[0] = tmp_comma + tmp_dec; 
        ret[1] = (float) humidity;
}

void Persistence::persistValue(float tmp, float hum, tm t) {
        if(!SPIFFS.begin(true)){
                Serial.println("An Error has occurred while mounting SPIFFS");
                return;
        }

        uint16_t year = t.tm_year + 1900;
        uint16_t month = t.tm_mon + 1;
        uint16_t day = t.tm_mday;
        uint16_t hour = t.tm_hour;
        uint16_t min = t.tm_min;

        uint16_t data_bytes = sensorDataToBytes(tmp, hum);

        File sensorData = SPIFFS.open(STORAGE_FILENAME, "ab+");
        sensorData.seek(0);

        if (!sensorData) {
                Serial.println("Failed to open Sensor-Datafile");
                return;
        }
        
        validateFileStructure(sensorData, year, YEAR_ID);
        validateFileStructure(sensorData, month, MONTH_ID);
        validateFileStructure(sensorData, day, DAY_ID);
        validateFileStructure(sensorData, hour, HOUR_ID);
        
        Serial.println("Validated File-Structure");

        while (true) {
                char line[5];
                int bytesRead = sensorData.readBytes(line, 5);
                 if (bytesRead == 5) {
                        uint16_t persistedTime = line[0] << 8 | line[1];
                        if (persistedTime == min) {
                                //Entry for curr minute already exists
                                break;
                        }
                } else if (bytesRead == 0) {
                        sensorData.write((uint8_t)(min >> 8));
                        sensorData.write((uint8_t)(min & 0b0000000011111111));
                        sensorData.write(':');
                        sensorData.write((uint8_t)(data_bytes >> 8));
                        sensorData.write((uint8_t)(data_bytes & 0b0000000011111111));
                        last_persisted_minute = min;
                        break;
                } else {
                        Serial.println("Wrong number of bytes read: " + bytesRead);
                        break;
                }
        }

        sensorData.close();
}

void Persistence::validateFileStructure(File sensorData, uint16_t curr_identifier_value, const char* id) {
        try {
                uint8_t line[5]; 
                while (true) { 
                        int bytesRead = sensorData.read(line, 5);
                        Serial.println(line[0]);
                        if (bytesRead == 5) {
                                if (line[0] == (uint8_t)id[0] && line[1] == (uint8_t)id[1]) {
                                        uint16_t a = line[3] << 8 | line[4];
                                        if (a == curr_identifier_value) {
                                            break;    
                                        }
                                }
                        } else if (bytesRead == 0) {
                                for (int i = 0; i < 3; i++) {
                                        sensorData.write((uint8_t)id[i]);
                                }

                                sensorData.write((uint8_t)(curr_identifier_value >> 8));
                                sensorData.write((uint8_t)(curr_identifier_value & 0b0000000011111111));

                                Serial.print("Created Entry: ");Serial.print(id[0]);Serial.println(id[1]);
                                break;
                        } else {
                                Serial.println("Wrong number of bytes read: " + bytesRead);
                                break;
                        }
                }
        } catch (std::exception& e) {
                Serial.println(e.what());
        }
}

float* Persistence::getPersistedDataByTimestamp(tm t, float* ret) {
        if(!SPIFFS.begin(true)){
                Serial.println("An Error has occurred while mounting SPIFFS");
                ret[0] = 0.0f;
                ret[1] = 0.0f; 
                return ret;
        }

        File sensorData = SPIFFS.open(STORAGE_FILENAME, FILE_READ);

        if (!sensorData) {
                Serial.println("Failed to open Sensor-Datafile");
                ret[0] = 0.0f;
                ret[1] = 0.0f; 
                return ret;
        }

        uint16_t year = t.tm_year + 1900;
        uint16_t month = t.tm_mon + 1;
        uint16_t day = t.tm_mday;
        uint16_t hour = t.tm_hour;

        uint16_t dataTime = t.tm_min;

        Serial.println(dataTime);

        uint8_t mtc[4][5] = {{YEAR_ID[0], YEAR_ID[1], YEAR_ID[2], year >> 8, year & 0b0000000011111111},
                             {MONTH_ID[0], MONTH_ID[1], MONTH_ID[2], month >> 8, month & 0b0000000011111111},
                             {DAY_ID[0], DAY_ID[1], DAY_ID[2], day >> 8, day & 0b0000000011111111},
                             {HOUR_ID[0], HOUR_ID[1], HOUR_ID[2], hour >> 8, hour & 0b0000000011111111}};
        
        uint8_t line[5] = {255, 255, 255, 255, 255};
        int currMtcIndex = 0;
        int bytesRead = sensorData.read(line, 5);
        while (bytesRead >= 0 || line[0] == 255) {
                if (currMtcIndex < 4) {
                        if (line[0] == mtc[currMtcIndex][0] && line[1] == mtc[currMtcIndex][1] && line[2] == mtc[currMtcIndex][2] 
                                && line[3] == mtc[currMtcIndex][3] && line[4] == mtc[currMtcIndex][4]) {
                                currMtcIndex++;
                        }
                } else {
                        uint16_t persistedMin = line[0] << 8 | line[1];
                        if (persistedMin == dataTime) {
                                break;
                        }
                }
                bytesRead = sensorData.read(line, 5);
        }

        sensorData.close();
        if (bytesRead > 0) {
                persistedBytesToValue(line[3], line[4], ret);
        } else {
                ret[0] = 0.0f;
                ret[1] = 0.0f;     
        }

        Serial.println(ret[0]);
        Serial.println(ret[1]);

        return ret;
}

std::string Persistence::decodeWholeFile() {
        if(!SPIFFS.begin(true)){
                Serial.println("An Error has occurred while mounting SPIFFS");
                return "An Error has occurred while mounting SPIFFS";
        }

        File sensorData = SPIFFS.open(STORAGE_FILENAME, FILE_READ);

        if (!sensorData) {
                Serial.println("Failed to open Sensor-Datafile");
                return "Failed to open Sensor-Datafile";
        }

        std::string body = "";

        uint8_t line[5]; 
        while (true) {
                int bytesRead = sensorData.read(line, 5);

                if (bytesRead == 5) {
                        if (line[0] == (uint8_t) 'y' || line[0] == (uint8_t) 'm' || line[0] == (uint8_t) 'd' || line[0] == (uint8_t) 'h') {
                                uint16_t val = line[3] << 8 | line[4];
                                body = body + (char) line[0] + (char) line[1] + (char) line[2] + intToString(val) + "\n";
                        } else {
                                uint16_t persistedMin = line[0] << 8 | line[1];
                                Serial.println(persistedMin);
                                float values[2];
                                persistedBytesToValue(line[3], line[4], values);

                                body = body + intToString(persistedMin) + (char) line[2] + floatToString(values[0]) + "," + floatToString(values[1]) + "\n";
                        }
                } else {
                        break;
                }
        }

        sensorData.close();

        Serial.println(body.c_str());
        return body;
}

void Persistence::overrideFile(std::string body) {
        if(!SPIFFS.begin(true)){
                Serial.println("An Error has occurred while mounting SPIFFS");
                return;
        }

        File sensorData = SPIFFS.open(STORAGE_FILENAME, FILE_WRITE);

        if (!sensorData) {
                Serial.println("Failed to open Sensor-Datafile");
                return;
        }

        size_t pos = 0;
        std::string line;
        while (pos != std::string::npos) {
                pos = body.find('\n');

                if (pos != std::string::npos) {
                        line = body.substr(0, pos - 1);
                        body.erase(0, pos+1);
                } else {
                        line = body;
                }

                int dp = line.find(':');
                std::string line_start = line.substr(0, dp);
                std::string line_end = line.substr(dp+1, line.size() -1);

                if (line_start == "yy" || line_start == "mm" || line_start == "dd" || line_start == "hh") {
                        uint16_t val;
                        std::istringstream iss(line_end); // Values 0 is not sucessfull
                        
                        iss >> val;

                        const char* pre = line_start.c_str();

                        sensorData.write((uint8_t) pre[0]);
                        sensorData.write((uint8_t) pre[1]);
                        sensorData.write((uint8_t) ':');
                        sensorData.write((uint8_t) (val >> 8));
                        sensorData.write((uint8_t) val & 0b0000000011111111);
                } else {
                        int tmp_sep = line_end.find(',');
                        std::string tmp_string = line_end.substr(0, tmp_sep);
                        int comma_pos = tmp_string.find('.');

                        std::string sdec = tmp_string.substr(0, comma_pos);
                        std::string scomma = tmp_string.substr(comma_pos + 1, 1);
                        std::string shum = line_end.substr(tmp_sep + 1, 2);

                        uint16_t tmp_dec;
                        uint16_t tmp_comma;
                        uint16_t hum_dec;
                        std::istringstream tdec(sdec);
                        std::istringstream tcomma(scomma);
                        std::istringstream thum(shum);

                        tdec >> tmp_dec;
                        tcomma >> tmp_comma;
                        thum >> hum_dec;

                        float tmp = tmp_dec + (((float) tmp_comma) / 10);
                        float hum = (float) hum_dec;

                        uint16_t ls;
                        std::istringstream iss(line_start);
                        iss >> ls;

                        uint16_t dataBytes = sensorDataToBytes(tmp, hum);

                        sensorData.write((uint8_t) (ls >> 8));
                        sensorData.write((uint8_t) (ls & 0b0000000011111111));
                        sensorData.write((uint8_t) ':');
                        sensorData.write((uint8_t) (dataBytes >> 8));
                        sensorData.write((uint8_t) (dataBytes & 0b0000000011111111));
                }       


        }

        sensorData.close();
        Serial.println("Overwritten sensorData.dat");
}

std::string Persistence::intToString(uint16_t value) {
        std::string res;
        while (value >= 10) {
                uint8_t number = value % 10;
                res = res + (char) (number + 48);
                value = value / 10;
        }

        res = res + (char) (value + 48);
        std::reverse(res.begin(), res.end());
        return res;
}

std::string Persistence::floatToString(float value) {
        std::string res_dec = intToString(floorf(value));

        float comma = value - floorf(value);
        uint8_t firstComma = floorf(comma * 10 + 0.5f);

        return res_dec + "." + (char) (firstComma + 48);
}